components/ai_chat/content/browser/ai_chat_tab_helper.cc

```diff
@@ -168,24 +135,25 @@ void AIChatTabHelper::WebContentsDestroyed() {
   inner_web_contents_ = nullptr;
 }
 
-void AIChatTabHelper::DidFinishNavigation(
-    content::NavigationHandle* navigation_handle) {
-  if (!navigation_handle->IsInMainFrame()) {
+void AIChatTabHelper::NavigationEntryCommitted(
+    const content::LoadCommittedDetails& load_details) {
+  if (!load_details.is_main_frame) {
     return;
   }
-  DVLOG(2) << __func__ << navigation_handle->GetNavigationId()
-           << " url: " << navigation_handle->GetURL().spec()
-           << " same document? " << navigation_handle->IsSameDocument();
-
+  // UniqueID will provide a consitent value for the entry when navigating
+  // through history, allowing us to re-join conversations and navigations.
+  int pending_navigation_id = load_details.entry->GetUniqueID();
+  pending_navigation_id_ = pending_navigation_id;
+  DVLOG(2) << __func__ << " id: " << pending_navigation_id_
+           << " url: " << load_details.entry->GetVirtualURL()
+           << " same document? " << load_details.is_same_document;
   // Allow same-document navigation, as content often changes as a result
   // of framgment / pushState / replaceState navigations.
   // Content won't be retrieved immediately and we don't have a similar
   // "DOM Content Loaded" event, so let's wait for something else such as
   // page title changing before committing to starting a new conversation
   // and treating it as a "fresh page".
-  is_same_document_navigation_ = navigation_handle->IsSameDocument();
-  pending_navigation_id_ = navigation_handle->GetNavigationId();
-
+  is_same_document_navigation_ = load_details.is_same_document;
```

Comment:
> also, fwiw I'm a little skeptical of the title change heuristic, but that's not for this PR 😆
It might not be perfect but I'd rather false negatives than false positives.

Although I did just notice a problem when testing - back navigations don't call `TitleWasSet`, they just update the title. The title is in `NavigationEntryCommitted`'s params though, so I can check if it's different there too. I'll have to add a test for this in the meatier history implementation PR.

> presumably the same document navigations will get a different id 
That's a good point. If I were to update the reference navigation ID without calling OnNewPage, same-document navigations which never result in a new conversation could get some bugs where results of async operations will be ignored because the navigation ID changed. And also the conversation would lose it's associated with the original navigation. Without doing that though, the conversation won't ever be associated with the current navigation when the user navigates back to it.
Seems like ideally we'd want it to be associated with multiple navigations. Should be simple enough as `AIChatService`'s map of navigation ID to conversation ID will be fine with the same conversation appearing multiple times.

Reply:
I have a "fix" for this on a separate branch - bf1781c5194077129c1fcbecdf7e75cf9d4a514a. That fix makes it so that the additional navigation IDs will be associated with the same conversation which will fix your scenario. I'm going to think more about it, and make a test, before I put it on this branch. It's not too neccessary for this branch since the conversation preservation and re-joining is behind a feature flag.

Reply:
Yup, happy to leave this for a followup PR